File: device-service/internal/service/application_error.go
package service

type DeviceApplicationError struct {
	Code    string
	Message string
}

func NewDeviceApplicationError(
	code string,
	message string,
) *DeviceApplicationError {
	return &DeviceApplicationError{
		Code:    code,
		Message: message,
	}
}

func (e *DeviceApplicationError) Error() string {
	return e.Message
}

File: frontend/src/application/core/applicationError.ts
export class ApplicationError extends Error {
  readonly code: string
  readonly causeValue?: unknown

  constructor(
    code: string,
    message: string,
    causeValue?: unknown,
  ) {
    super(message)
    this.name = "ApplicationError"
    this.code = code
    this.causeValue = causeValue
  }
}
